### ISSUE TITLE
Replay HTTP requests against an air-gapped Confluence server

### ISSUE THREAD

**Issue by oxbqkwwxfrqccwtg** [sphinx-contrib/confluencebuilder#823]
Hi all,

i am currently facing the situation where i need to **publish to an air-gapped Confluence server inside a virtualised and privatised environment (over Windows VDI)**. In addition, company policy forbids me from using, or installing Python on the VDI. I can freeload a perl executable that came bundled with Git for Windows, but there are no other scripting means besides Windows PowerShell.

So... I'm either thinking about hacking something in PowerShell or Perl.

## Quick-n-Nasty Approach

1. Figure out what data *confluencebuilder* is pulling from the target server and what data is used in the confluencebuilder output
2. write a small pass-through wrapper for all HTTP calls made against the bundled `requests` package. (i'd use `unittest.mock` or something similar)
2. index all POST, PUT, etc. HTTP request objects and write a generic replay sequence
3. Replay all HTTP requests against the air-gapped Confluence server

## Quick-n-Dirty Approach

Basically just like the Quick-n-Nasty solution except that i will rewrite the `ConfluencePublisher` class

I just require this as an interims workaround, since i'll be getting access to proper CI/CD facilities soon.

## My question

Was a scenario, as described by me, ever considered during the development? I'd be interested in the thoughts.

It's probably not worth adding it as a feature, but *publishing replays* could be a feature, if the publishing task was split into building and publishing, where the build task could output an index of HTTP requests metadata and request bodies and the publisher executes upon the index.

**Comment by jdknight** [comment]
This topic was discussed in https://github.com/sphinx-contrib/confluencebuilder/issues/627 last year, if you would like to read up about the opinions/issues at the time.

Note that part of the hesitation outlined in these comments were due to some prospect implementation in supporting Confluence' ADF format (in part to support the v2 editor style). The implementation would require a multi-pass publishing design, otherwise Confluence would reject uploads. Since this extension was able to support the v2 editor while still using the storage format, publishing should be as simple as it is right now in the current implementation. So, when considering the other issue's comments, it may be somewhat easier to implement a separation between building and publishing.

While my first impressions think that a replay functionality should not be something directly implemented in this extension, I do think it's a great idea for your use case. Isolating request data made should be easily translatable to PowerShell (or whatever) scripts. While the suggested approach of wrapping on the requests library should work, I wonder if it would be simpler if this extension supported an advanced configuration or introduce extension hooks which allows a user to override either the `ConfluencePublisher` or `Rest` classes without needing to use mock implementation (but I guess it doesn't really matter what is used; whatever works for this use case). This Confluence builder extension is always willing to create hooks or support overrides from the configuration for existing features, if power users want an easier way to perform custom tweaks.

The biggest thing to consider when looking at the separation between building and publishing is not so much the generated documentation set, but the dynamic values needed in some of the data requests. We could initially ignore some features like purging old pages or publishing on updated pages/attachments, and just focus on publishing an entire documentation/attachment set. One example to consider is that the `version`'s [`number`](https://github.com/sphinx-contrib/confluencebuilder/blob/v2.1.1/sphinxcontrib/confluencebuilder/publisher.py#L1118), which on new pages is set to one (or not provided). However, a publish attempt is updating an existing document, this value must be set to the existing version number of the page [plus one](https://github.com/sphinx-contrib/confluencebuilder/blob/v2.1.1/sphinxcontrib/confluencebuilder/publisher.py#L1162). This can make it hard if you are trying to do a dumb-replay publish ~ you may have to add some smarts to know a request needs an update and tweaks the version value. Alternatively, you could also just crudely wipe the entire space and publish a new set with all versions being `1`.

As for the Confluence Builder extension, my current opinion on supporting the separation of building and publishing is as follows. I am in favor to add support for such a feature. I would not say it is a priority feature to get out by next release, but it could be the next big feature we aim to bring in this extension. The following overview is how I would imagine such a design would be done:

```mermaid
flowchart LR
    P --> S
    Tpp --> S
    E --> A
    A -.-> Tpp
    A -.-> I
    subgraph Sphinx/Confluence Builder
    C[Initialization] --> B
    C --> I
    B[Builder] --> P
    B --> E
    E[Export]
    I[Import] --> P
    P[Publisher]
    end
    A[Archive]
    subgraph Third Party
    Tpp[Publisher]
    end
    subgraph "Confluence"
    S[Space]
    end
```

I'm welcome to other design considerations as well.

Finally, I would imagine the timelines wanted for OPs requirements may not be align with the timelines of this extension providing full support for build/publish separation. For sure it should be possible to tweak this extension in order to be flexible for a third-party replay functionality (if tweaks to this extension are needed), but a serious look at a full fledged solution may not be looked into (by this maintainer) until possible Fall, if not next year.

**Comment by tiara-adessi** [comment]
Marry me! 🥹 Thank you so much for such a quick and precise reply!

> Alternatively, you could also just crudely wipe the entire space and publish a new set with all versions being 1.

That's exactly what i had in mind. 😄 Quick-n-Nasty!

Should this become a feature and it will be implemented similar to what @jdknight proposes, i'd volunteer for providing a reference implementation of a third-party publisher. Having a plan B is always nice, so i don't see a problem in me dedicating time for this, probably Q4 '23.

**Comment by tiara-adessi** [comment]
Attached you'll find my quick-n-dirty boilerplate implementation of the wrapper. Fixing up a custom Rest Client in PowerShell/Perl/whatever can be achieved in less than 50 lines, hence i don't see the point in sharing.

Maybe this will be useful for someone else.

@jdknight if i can support the project with any feature implementations, let me know. I feel comfortable with the code now.

In addition, i've appended a schema for the output. 

By default the dump will go to `sphinx.config.Config.outdir` / 'confluence.out'.

One can use the schema in conjunction with e.g. the `TestJSON` PowerShell cmdlet.

Attachment data is decoded as *ISO-8859-1*.

```python
#!/usr/bin/env python3
"""Publishment delay wrapper sphinxcontrib.confluencebuilder

This is a lightweight pass-through wrapper for 
``sphinxcontrib.confluencebuilder``, which intercepts all ``store_*`` calls on 
a ``ConfluencePublisher`` instance, dumps all data into interchange. 
The index and dumps can be used in conjunction with the PowerShell 
helper to delay/replay the publishment of pages and attachments for a different 
Confluence instance, than what the programmatic target is.

The builder name is ``x_confluence``

The use-case for this implementation is as follows:

    I am currently facing the situation where i need to publish to an 
    air-gapped Confluence server inside a virtualised and privatised 
    environment (over Windows VDI). In addition, company policy forbids me from 
    using, or installing Python 
    on the VDI. I can freeload a perl executable that came bundled with Git for 
    Windows, but there are no other scripting means besides Windows PowerShell.
"""
__author__ = 'tiara.rodney@adesso.de'
__copyright__ = 'DL-DE-BY-2.0'

from dataclasses import dataclass, asdict
import json
from pathlib import Path
from typing import Any, Optional, ByteString, List
from unittest.mock import patch, Mock

from sphinx.application import Sphinx
from sphinx.util import logging

from sphinxcontrib.confluencebuilder import setup as _setup
from sphinxcontrib.confluencebuilder.builder import (
    ConfluenceBuilder as _ConfluenceBuilder
)
from sphinxcontrib.confluencebuilder.publisher import (
    ConfluencePublisher as _ConfluencePublisher
)


logger = logging.getLogger(__name__)


@dataclass
class ConfluencePublisherPage:
    """
    """
    #:
    page_name: str

    #: 
    page_id: str

    #: 
    parent_id: str

    #:
    data: dict


@dataclass
class ConfluencePublisherAttachment:
    """
    """
    #: 
    page_id: str

    #: 
    name: str

    #: 
    data: ByteString

    #: 
    mimetype: str

    #: 
    attachment_id: str


@dataclass
class ConfluencePublisherDump:
    """
    """
    #: 
    pages: List[ConfluencePublisherPage]

    #: 
    attachments: List[ConfluencePublisherAttachment]


class ConfluencePublisher(_ConfluencePublisher):
    """
    """
    def __init__(self):
        """
        """
        super().__init__()

        self.dump = ConfluencePublisherDump(
            pages = [],
            attachments = []
        )

    def __getattr__(self, name: str) -> Any:
        """
        """
        return super().__getattribute__(name)


    def __setattr__(self, name: str, value: Any) -> None:
        """
        """
        return super().__setattr__(name, value)

    def store_attachment(
        self,
        page_id: str,
        name: str,
        data: Any,
        mimetype: Any,
        hash_: str,
        force: bool = False
    ) -> str:
        """request to store an attachment on a provided page

        :returns: the attachment identifier
        """
        logger.info('pass-through intercept: store_attachment')

        logger.info(getattr(self.config, 'x_confluence_outdir'))

        result = super().store_attachment(page_id, name, data, mimetype, hash_,
                                          force)

        self.dump.attachments.append(ConfluencePublisherAttachment(
            page_id = page_id,
            name = name,
            data = data.decode('ISO-8859-1'),
            mimetype= mimetype,
            attachment_id = result
        ))

    def store_page(
        self,
        page_name: str,
        data: Any,
        parent_id: Optional[str] = None
    ) -> str:
        """request to store page information to a confluence instance

        :param page_name: the page title to use on the updated page
        :param data:  the page data to apply
        :param parent_id: the id of the ancestor to use

        :returns: id of uploaded page
        """
        logger.info('pass-through intercept: store_page')

        logger.info(parent_id)

        result = super().store_page(page_name, data, parent_id)

        self.dump.pages.append(ConfluencePublisherPage(
            page_name = page_name,
            data = data,
            parent_id = parent_id,
            page_id = result
        ))

        return result

    def store_page_by_id(
        self,
        page_name: str,
        page_id: str,
        data: Any
    ) -> str:
        """request to store page information on the page with a matching id

        :param page_name: the page title to use on the updated page
        :param data:  the page data to apply
        :param parent_id: the id of the ancestor to use

        :returns: id of uploaded page
        """
        logger.info('pass-through intercept: store_page_by_id')

        return super().store_page_by_id(page_name, page_id, data)


    def disconnect(self):
        """
        """
        file = Path(getattr(self.config, 'x_confluence_outdir')) / 'data.json'

        file.parent.mkdir(parents = True, exist_ok=True)

        raw = json.dumps(asdict(self.dump), indent=4)

        file.write_text(raw)

        return super().disconnect()


class ConfluenceBuilder(_ConfluenceBuilder):
    """
    """
    name = 'x_confluence'

    def __init__(self, app: Sphinx, env = None):
        """
        """
        patch_target = ('sphinxcontrib.confluencebuilder'
                        '.builder.ConfluencePublisher')

        with patch(patch_target, ConfluencePublisher):

            super().__init__(app, env)


    def __getattribute__(self, name: str) -> Any:
        """
        """
        return super().__getattribute__(name)


    def __setattr__(self, name: str, value: Any) -> None:
        """
        """
        return super().__setattr__(name, value)


def setup(app: Sphinx):
    """
    """
    patch_target = 'sphinxcontrib.confluencebuilder.ConfluenceBuilder'

    app.add_config_value(
        name = 'x_confluence_outdir',
        default = str(Path(app.outdir) / 'confluence.out'),
        rebuild = True
    )

    with patch(patch_target, ConfluenceBuilder):

        logger.info(f'patching: {patch_target}')

        return _setup(app)
```

```python
{
	"$id": "https://github.com/tiara-adessi/confluencebuilder/schema/top",
    "x-authors": [
        "tiara.rodney@adesso.de"
    ],
    "type": "object",
    "properties": {
        "pages": {
            "type": "array",
            "items": {
                "$ref": "#/definitions/page"
            }
        },
        "attachments": {
            "type": "array",
            "items": {
                "$ref": "#/definitions/attachment"
            }
        }
    },
    "required": [
        "pages",
        "attachments"
    ],
    "definitions": {
        "page": {
            "type": "object",
            "properties": {
                "page_name": {
                    "type": "string"
                },
                "page_id": {
                    "type": "string"
                },
                "parent_id": {
                    "type": "string"
                },
                "data": {
                    "type": "object",
                    "properties": {
                        "content": {
                            "type": "string"
                        },
                        "labels": {
                            "type": "array",
                            "items": {
                                "type": "string"
                            }
                        }
                    },
                    "required": [
                        "content",
                        "labels"
                    ]
                }
            },
            "required": [
                "page_name",
                "page_id",
                "parent_id",
                "data"
            ]
        },
        "attachment": {
            "type": "object",
            "properties": {
                "page_id": {
                    "type": "string"
                },
                "name": {
                    "type": "string"
                },
                "data": {
                    "type": "string"
                },
                "mimetype": {
                    "type": "string"
                },
                "attachment_id": {
                    "type": "string"
                }
            },
            "required": [
                "page_id",
                "name",
                "data",
                "mimetype",
                "attachment_id"
            ]
        }
    }
}
```

**Comment by jdknight** [comment]
Thanks for the example, provides some good insights to understand what is desired.

I'm curious about the schema definition, since I cannot say I've created a schema definition file for JSON before. Noticed there is the use of `type`, `properties`, `required`, etc. which looks to be able to support a way to programmatically understand the schema definition. Is this based off a common practice (e.g. using these specific keys)? Manually created or was this created with something else?

**Comment by tiara-adessi** [comment]
That was a very nice subtle hint, that i was missing the [`$schema`](https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-01#name-the-schema-keyword) keyword (though it's not mandatory). 😄

Yes, these are common practices as defined in the JSONSchema specification(s).

 There are some pretty good validators out there (e.g., I use [this](https://pypi.org/project/jsonschema/) one in Python, and [this](https://www.npmjs.com/package/jsonschema) one in Node.js, and [this](https://metacpan.org/pod/JSON::Schema) one in Perl).
The nice thing about Microsoft PowerShell is, that they have a built-in [Newtonsoft](https://www.newtonsoft.com/json) validator through the [`Test-JSON`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/test-json?view=powershell-7.3) cmdlet, which is available for PowerShell Core and Desktop. So Microsoft seems to be leading the way into making JSONSchema actually more practical (it's stuck in the *forever-draft* dimension)

Yes, the schema was created manually and i tend to use older versions of the specification to author them, so that it stays compatible across multiple validators.

Besides the [`$ref`]() reference resolution (the specification is vague about that, therefore there are multiple understandings and implementation approaches of that) it is pretty straight forward.

**Comment by jdknight** [comment]
Over a year later....

The development branch now includes support for generating a manifest file in a project's output directory. Manifests are always generated (we can change this later if users report any issues with such a generation).

The proposed information and sample schema has been helpful; much appreciated. The manifest schema used does vary in some keys, but it is believed it captures the original proposed information (and more). Feel free to inspect a sample output in the introducing pull request (https://github.com/sphinx-contrib/confluencebuilder/pull/1085).

The change also supports a [`confluence_manifest_data`](https://sphinxcontrib-confluencebuilder.readthedocs.io/en/latest/configuration/#confval-confluence_manifest_data) configuration, which allows also page/attachment data to be embedded into the manifest. This might prove useful for tooling that does not want to manage any specific files if wanting to publish pages/attachments (i.e. it can used the Base64-encoded content instead).

Just to clarify, this does not provide a complete solution for a two-pass publishing support. The manifest file should be enough to allow a user to use custom tooling to perform a fresh publish event for environments like an air-gapped environment. However, this extension does not provide a means to process a manifest file and publish it. The publishing aspect of this extension is heavily driven on querying a Confluence instance to help generate required page/attachment/etc. update requests (e.g. version bumping page updates); somewhat already covered above in other comments. Trying to support a perfect two-pass setup is not an aim of this extension's capabilities. Going forward, this extension will be willing to add/maintain manifest support which external tooling can use.

With that being said, support for manifests are available now with a [development installation](https://sphinxcontrib-confluencebuilder.readthedocs.io/install/#development-installation) and should be available next stable release. Once available in a release, this issue will planned to be closed. Feel free to comment and mention anything that might be missing in this effort.

**Comment by jdknight** [comment]
v2.10 is now available on PyPI -- marking as closed.

**Comment by oxbqkwwxfrqccwtg** [comment]
@jdknight Very, very nice! :) I will take it for a test drive soon.

**Comment by oxbqkwwxfrqccwtg** [comment]
@jdknight, thanks for having considered this as a feature.

I'm planning on refactoring the [PowerShell publisher](https://bitbucket.org/tiaracodes/psconfluencepublisher) I've originally implemented, prior to this feature existing. Would it be of use to serve as a reference implementation? If so, do you have any requirements towards how I should document it?

**Comment by jdknight** [comment]
> I'm planning on refactoring the PowerShell publisher...

Nice!

> Would it be of use to serve as a reference implementation? If so, do you have any requirements towards how I should document it?

I'm not entirely sure what the specific ask is here. My initial impression is to code/document the PSConfluencePublisher application as you see fit. If you update it in a way where it works with this extension's manifests, we could probably add something to the sphinxcontrib-confluencebuilder extension's documentation (maybe under a new section "Features -> Manifest", after "Extensions" but before "Other") which provides a generic description of what the manifests are, add a section indicating tools/implementations that support using this manifest and then reference PSConfluencePublisher's home.

Please let me know if I'm overlooking something here.

**Comment by oxbqkwwxfrqccwtg** [comment]
That answers my question, thank you. I'll open a new issue, once I've completed the refactoring and documentation.
